Smoke testing with Cucumber

radtriste opened the pull request:
https://issues.redhat.com/browse/KOGITO-491

Go implementation of Cucumber tests, using [godog](https://github.com/DATA-DOG/godog) official framework.

ricardozanini commented on test/smoke/go/kube.go:
```diff
@@ -0,0 +1,130 @@
+)
+
+const (
+	ENV_VAR_KUBECONFIG = "KUBECONFIG"
```
Same thing here regarding `const` names

ricardozanini commented on test/smoke/go/kube.go:
```diff
@@ -0,0 +1,130 @@
+		if err != nil {
+			panic(fmt.Sprintf("Error building Kube connection config: %v", err))
+		}
+		kubeClient.BuildCli, err = clientbuildv1.NewForConfig(config)
```
I guess the `BuildCli` could be build in the `NewForConsole` function or a new function in that package, like `client.WithBuildConfig().NewForConsole()`

radtriste replied:
Made some modifications on that part, I let you review it.

we can now do `client.newBuilder().With*().Build()`
I updated also `NewForConsole` and `NewForController` with this form.

ricardozanini commented on test/smoke/go/kube.go:
```diff
@@ -0,0 +1,130 @@
+	return config, nil
+}
+
+func getKubeConfigFile() string {
```
I think we already have a similar function:
https://github.com/kiegroup/kogito-cloud-operator/blob/master/pkg/client/client.go#L183

radtriste replied:
Solved as I created the builder and so it reuses the code from client

ricardozanini commented on test/smoke/go/kube.go:
```diff
@@ -0,0 +1,130 @@
+	return kubeconfig
+}
+
+func loadResource(namespace string, uri string, resourceRef meta.ResourceObject, beforeCreate func(object interface{})) error{
```
How about that?
https://github.com/kiegroup/kogito-cloud-operator/blob/master/cmd/kogito/command/shared/install_operator.go#L126

We could refactor into a similar functionality

radtriste replied:
I refactored the part loading the yaml file, creating a `createFromYamlContent` method on resource from client. Please let me know what you think of it.

ricardozanini commented on test/smoke/go/operator.go:
```diff
@@ -0,0 +1,129 @@
+	return true, nil
+}
+
+func GetOperatorDeployment(namespace string) (*coreapps.Deployment, error) {
```
We have this:
https://github.com/kiegroup/kogito-cloud-operator/blob/1e89c86ddc2e28728cffa3b78f899d2c6e110170/cmd/kogito/command/shared/resource_checks.go#L70

radtriste replied:
Reused it. Was just not a big fan reusing some logic from cli or operator controller.
I reuse client which is more or less generic and could be reused in other projects
I do need the apis and a change in it, I will see it really fast.
Reuse the logic can be risky IMO. But for now I reuse it.

radtriste replied:
or we would need to create generic package with those methods.

ricardozanini replied:
We have to move this function to a shared package. I suggest `pkg/infrastructure`.

spolti commented on pkg/client/openshift/buildconfig.go:
```diff
@@ -145,7 +145,6 @@ func (b *buildConfig) GetBuildsStatus(bc *buildv1.BuildConfig, labelSelector str
 
 func (b *buildConfig) checkBuildConfigExists(bc *buildv1.BuildConfig) (bool, error) {
 	if _, err := b.client.BuildCli.BuildConfigs(bc.Namespace).Get(bc.Name, metav1.GetOptions{}); err != nil && errors.IsNotFound(err) {
-		log.Warnf("BuildConfig not found in namespace")
```
+1

radtriste replied:
Set back

spolti commented on pkg/controller/kogitodataindex/resource/testdata/onboarding-dockerimage-json:
```diff
@@ -36,7 +36,7 @@
             "while [ ! -f \"/tmp/s2i-upload-done426790037\" ]; do sleep 0.5; done; /usr/local/s2i/assemble-runtime; exit $?"
         ],
-        "Image": "quay.io/kiegroup/kogito-quarkus-jvm-ubi8:0.5.0",
+        "Image": "quay.io/kiegroup/kogito-quarkus-jvm-ubi8:0.6.0",
```
maybe consider to use 0.6.1, it will be ready by Monday.

radtriste replied:
I rebased from master and did not change any version from it. It should be directly update or should I do it here ? (but who knows when it will be merged ...)

spolti commented on test/smoke/go/kogitoapp.go:
```diff
@@ -0,0 +1,85 @@
+)
+
+func DeployNonNativeQuarkusExample(namespace, appName, contextDir string) error{
+	fmt.Println("Deploy non native quarkus example " + contextDir + " with name " + appName)
```
can't we import the logging here?

radtriste replied:
I used the logger package in the test files

spolti commented on test/smoke/go/operator.go:
```diff
@@ -0,0 +1,129 @@
+)
+
+const (
+	DEFAULT_OPERATOR_IMAGE="quay.io/kiegroup/kogito-cloud-operator"
```
indentation looks wrong.
plus the const pattern.

radtriste replied:
corrected

spolti commented on test/smoke/java/smoke-tests/src/main/java/org/kie/kogito/operator/tests/ops/BasicOpenshiftOperations.java:
```diff
@@ -0,0 +1,198 @@
+        }
+    }
+
+    private String retrievePort(Protocol protocol) {
```
wouldn't be better create a different class and put the enum plus this method there?

radtriste replied:
Updated the enum

spolti commented on version/version.go:
```diff
@@ -16,5 +16,5 @@ package version
 var (
 	// Version of Kogito Operator
-	Version = "0.7.0-rc1"
+	Version = "0.6.0"
```
shouldn't be updated.

radtriste replied:
Set back

spolti commented on test/smoke/java/smoke-tests/src/test/java/org/kie/kogito/operator/tests/RunCucumberTest.java:
```diff
@@ -0,0 +1,30 @@
+import io.cucumber.testng.CucumberOptions;
+import org.testng.annotations.DataProvider;
+
+//@RunWith(Cucumber.class)
```
why is it commented out?

radtriste replied:
It was for JUnit but as it uses testNG, I deleted it

radtriste commented on go.mod:
```diff
@@ -36,7 +36,7 @@ replace (
 
 // thrift moved to github
-replace git.apache.org/thrift.git => github.com/apache/thrift v0.12.0
+// replace git.apache.org/thrift.git => github.com/apache/thrift v0.12.0
```
This is for m local testing. Forgot to remove it ... good catch !

radtriste replied:
Set back

radtriste commented on cmd/kogito/version/version.go:
```diff
@@ -16,5 +16,5 @@ package version
 var (
 	// Version of Kogito CLI
-	Version = "0.7.0-rc1"
+	Version = "0.6.0"
```
Set back

radtriste commented:
Updated Go code from feedbacks. 
Not sure we need to keep the Java part. @sutaakar @jakubschwan wdyt ?

Also added some test configuration on Go part, see README.md

ricardozanini commented on go.mod:
```diff
@@ -1,6 +1,7 @@
 module github.com/kiegroup/kogito-cloud-operator
 
 require (
+	github.com/DATA-DOG/godog v0.7.13
```
Why do we need this dependency?

radtriste replied:
This is the cucumber dependency for Go

ricardozanini commented on cmd/kogito/command/shared/resource_checks.go:
```diff
@@ -67,7 +67,7 @@ func CheckProjectExists(kubeCli *client.Client, namespace string) error {
 }
 
-func checkKogitoOperatorExists(kubeCli *client.Client, namespace string) (bool, error) {
+func CheckKogitoOperatorExists(kubeCli *client.Client, namespace string) (bool, error) {
```
This function should moved to `pkg/infrastructure`

radtriste replied:
operator go file created into pkg/infrastructure

ricardozanini commented on pkg/client/client_builder.go:
```diff
@@ -0,0 +1,136 @@
+}
+
+// Builder wraps information about what to create for a client before building it
+type Builder struct {
```
Create a `builder` struct (turn this into a private struct) and an Interface named `Builder` (public), then the private one will implement the public. This way we hide the implementation from external packages and follow the Golang standard. 

Nice implementation! Congrats!

radtriste replied:
done

ricardozanini commented on pkg/client/kubernetes/resource.go:
```diff
@@ -38,6 +43,7 @@ type ResourceInterface interface {
 	Delete(resource meta.ResourceObject) error
 	UpdateStatus(resource meta.ResourceObject) error
 	Update(resource meta.ResourceObject) error
+	CreateFromYamlContent(yamlContent, namespace string, resourceRef meta.ResourceObject, beforeCreate func(object interface{})) error
```
Could you please add the documentation in the interface instead?

radtriste replied:
You mean just right here ?

ricardozanini commented on test/smoke/go/operator.go:
```diff
@@ -0,0 +1,120 @@
+	rbac "k8s.io/api/rbac/v1"
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	
+	"github.com/kiegroup/kogito-cloud-operator/cmd/kogito/command/shared"
```
Let's not import from `cmd`

radtriste replied:
import now pkg/infrastructure

sutaakar commented:
I think we can get rid of Java part.

ricardozanini commented on cmd/kogito/command/shared/install_operator.go:
```diff
@@ -66,7 +65,7 @@ func MustInstallOperatorIfNotExists(namespace string, operatorImage string, clie
 	}
 
-	if exists, err := checkKogitoOperatorExists(client, namespace); err != nil {
+	if exists, err := infra.CheckKogitoOperatorExists(client, namespace); err != nil {
```
`infra` in French is a short cut for infrastructure? It's in Portuguese as well, but IDK if it's in English also. I'm just curious because I'm always tempting to use `infra` all the time instead. :p

ricardozanini commented on hack/run-smoke.sh:
```diff
@@ -0,0 +1,74 @@
+# runs all smoke tests for the operator
+SCRIPT_NAME=`basename $0`
+
+function usage(){
```
we could make this a default option for all scripts in this dir.

radtriste replied:
We should open an issue to update the scripts.

ricardozanini commented on pkg/client/client_builder.go:
```diff
@@ -0,0 +1,153 @@
+
+// Builder wraps information about what to create for a client before building it
+type Builder interface {
+	UseConfig(kubeconfig *restclient.Config) Builder
```
The method documentation must be into the interface instead of in the implementation.

sutaakar commented on test/smoke/main_test.go:
```diff
@@ -158,7 +161,7 @@ func FeatureContext(s *godog.Suite) {
 
 	s.BeforeScenario(data.setUp)
-	s.AfterScenario(data.tearDown)
+	//s.AfterScenario(data.tearDown)
```
Leftover from debugging?

sutaakar commented on test/smoke/wip/deploy_quarkus_with_persistence.feature:
```diff
@@ -7,14 +7,16 @@ Feature: Service Deployment: Quarkus with persistence
+    And HTTP POST request on service "jbpm-quarkus-example" with path "orders" and "json" body '{"approver" : "john", "order" : {"orderNumber" : "12345", "shipped" : false}}'
+    And HTTP GET request on service "jbpm-quarkus-example" with path "orders" should return an array of size 1 within 1 minutes
     
     When Scale DeploymentConfig "jbpm-quarkus-example" to 0 pods within 2 minutes
```
This is not reliable, if operator detects that DC has been scaled manually it will override DC. If operator manage to do it before OpenShift detects that DC has been scaled then respin won't happen.

I think that it would be better to scale DC by setting replicas parameter in CR.

radtriste commented:
@sutaakar @ricardozanini Could you please review last changes ?
If you find nothing more, PR is ready to be merged ;)

sutaakar commented on test/smoke/kogitoapp.go:
```diff
@@ -0,0 +1,141 @@
+)
+
+const (
+	//kogitoExamplesURI = "https://github.com/kiegroup/kogito-examples"
```
Why is this value commented out?

radtriste replied:
I needed to use my own fork of the kogito-examples as the kiegroup quarkus examples are missing the infinispan application.properties for persistence ...

radtriste replied:
@ricardozanini wasn't it supposed to be corrected if properties are missing ?

sutaakar replied:
I think this should be fixed in Quarkus 1.1.0?

sutaakar replied:
Anyway, please add a comment there, explaining the reasons for substitution, so it can be reverted later.

sutaakar commented on test/smoke/kogitoapp.go:
```diff
@@ -0,0 +1,143 @@
+	kogitoApp.Spec.Build.Native = native
+	kogitoApp.Spec.Build.GitSource.URI = &gitProjectURI
+	kogitoApp.Spec.Build.GitSource.ContextDir = contextDir
+	kogitoApp.Spec.Build.GitSource.Reference = "0.6.1"
```
Why there is used specific branch and not master?
If specific branch needs to be used, can it be configurable? For example using env variable?

radtriste replied:
Branch master had problems was not compiling ... :S
I will set this one and the URI as configurable via env variable.

radtriste commented:
@sutaakar let me know

radtriste commented:
sure

sutaakar commented:
@radtriste Looks good now

radtriste commented:
@spolti @sutaakar (somehow your approval is not there ...) @msmagnanijr @jakubschwan @xiezhang7 
could we have approval on that ? so we can merge it today

spolti commented:
Will this smoke test be part of our CI? 
@radtriste @ricardozanini maybe add a GA Actions step to see if it would work well?

radtriste commented:
@spolti We still have the problem of run the tests on an OpenShift

spolti commented:
@radtriste Sure, np. Thanks!

radtriste commented on README.md:
```diff
@@ -829,6 +831,44 @@ where:
 
+### Running Smoke tests
+
+If you have an OpenShift cluster and admin privileges, you can run Smoke tests with the following command:
```
updated
